### PR TITLE
Only add direct neighbors for each MapBlock loaded on the client.

### PR DESCRIPTION
**Pull request by lhofhansl**
While profiling only on the main Minetest thread (the one that also renders the view) I noticed that when I move around on the map VoxelManipulator::copyFrom and memcpy/memset called from it are using a large percentage of the CPU as blocks are loaded into the client.

I eventually tracked it down to MeshMakeData::fill. If I understand the intention of the code correctly the goal is to make sure all border meshes are rendered eventually by making sure all neighboring blocks are loaded.

In order to achieve that it is not necessary to also (re-)add all 26 neighbors to the VoxelManipulator but only the 6 that actually share a "surface".

(To verify when I remove that part altogether I noticed many missing surfaces. When I change the code as suggested in this PR there are no rendering problems even after prolonged checking with many scenarios)

This cut down the the time spent in copyFrom from > 17% to < 5%.

### CODE REVIEW

**Comment by paramat** [comment]
Nice simple PR. Perhaps i'll try to catch c55 and ask him if there was a reason for 26 instead of 6.

File: src/mapblock_mesh.cpp [hunk 1]
```diff
@@ -83,9 +83,9 @@ void MeshMakeData::fill(MapBlock *block)
 		// Get map
 		Map *map = block->getParent();
 
-		for(u16 i=0; i<26; i++)
```
Comment (rubenwardy) [inline src/mapblock_mesh.cpp]: also correct this code style

Comment (lhofhansl) [inline src/mapblock_mesh.cpp]: done

File: src/mapblock_mesh.cpp [hunk 2]
```diff
@@ -83,9 +83,8 @@ void MeshMakeData::fill(MapBlock *block)
...
-		{
-			const v3s16 &dir = g_26dirs[i];
+		for (u16 i=0; i<6; i++) {
```
Comment (paramat) [inline src/mapblock_mesh.cpp]: Spaces around '=' and '<'

**Comment by paramat** [comment]
Looks like there is a reason https://github.com/minetest/minetest/commit/5c26972cddb78f3f93d5d6e0428a6b282fde4bfe

**Comment by paramat** [comment]
c55 > "meshes at the corners depend on those"

**Comment by lhofhansl** [comment]
Hmm... Shame. This is very wasteful.

I tried more, I cannot see any rendering issues with this change.

Under which circumstances would this lead to any issues?

**Comment by paramat** [comment]
Maybe we can ask c55 again later, that's all he has said so far.

**Comment by Ekdohibs** [comment]
However, why do we need to copy the whole mapblocks and not just a single node of each of them? Wouldn't that improve performance even more?

**Comment by sfan5** [comment]
@Ekdohibs 
see the comment above: https://github.com/minetest/minetest/commit/5c26972cddb78f3f93d5d6e0428a6b282fde4bfe#diff-c2f90db5de8c33a259c27113939c63c5R73

**Comment by Ekdohibs** [comment]
Interesting, would be worth looking at some way to avoid copying all this data then (that's roughly 300KB of data we're copying each time...)

**Comment by paramat** [comment]
I've askes c55 to describe a situation that caused glitches when he made that commit.

**Comment by celeron55** [comment]
http://irc.minetest.net/minetest-dev/2012-11-27
https://github.com/minetest/minetest/issues/307

**Comment by paramat** [comment]
Since there have been some changes to lighting it's worth checking again,

**Comment by Fixer-007** [comment]
With PR:
with
Without PR:
without

Look diagonal block near torch

**Comment by paramat** [comment]
Should be checked at mapchunk corners too as that is apparently more extreme.
Looks like the new light curve might have improved how this bug looks.

**Comment by Fixer-007** [comment]
Added screenshot for edge... ^^

**Comment by lhofhansl** [comment]
Fair enough. Let's close this one.

The profiler did flag this one as the main CPU drain on the main thread when blocks are loaded.

Fps would not get better when everything is loaded. What should improve is the jitter when new blocks are loaded. In my case significantly.

**Comment by lhofhansl** [comment]
Reduces the typical jitter from 100-200% to 60-70% when I move into a new map. Would be nice if somebody could confirm, all you need to do is fly into a new area of the map.

@Fixer-007 I've seen you using some tool to produce repeatable movements through maps, what's the tool you've been using there.

(please note my time for this is fairly limited, which is why I've been doing only little fixes lately - and not quite as successful on those either)

**Comment by lhofhansl** [comment]
Ahh. It was FRAPS. Windows only, so I cannot use it here.

**Comment by celeron55** [comment]
If this considerably helps with jitter on some systems, situations, builds or whatnot, then this can be considered to be added as an option. Does someone have opinions about that?

**Comment by lhofhansl** [comment]
There's also this: [client.cpp#L1602](https://github.com/minetest/minetest/blob/master/src/client.cpp#L1602
), which also pulls in the 6 direct neighbors. All in all quite wasteful, in this case we're copying 7*27=189 map blocks.
Not sure why this is needed here as well, but at the very least we could avoid the duplicate copies and bring in the 5^3 map block volume (125 map blocks, avoiding copying 64 map blocks at least).

**Comment by juhdanad** [comment]
Some nodes need to know their 18 neighbors (neighbors that have at least a common edge). With loading only 6 neighbor blocks, they look bad:
framed glass and rail:
screenshot_20170217_122629
water:
screenshot_20170217_122644

**Comment by Fixer-007** [comment]
>Ahh. It was FRAPS. Windows only, so I cannot use it here.

Fraps + tinytask with most stuff turned off in background, I did just 1 run again with this PR, and I just don't see reduction in stutters... looks mostly the same :(

**Comment by Fixer-007** [comment]
If you want to look more into stuttering, I have smth interesting for you. 
Minetest 0.2.20110602 does not have stutter at all on my PC, yet minetest-0.2.20110731_3 (https://github.com/minetest/minetest/tree/0.2.20110731_3) version does, thats just 1.5 month difference, if you can look what was changed and what started it all, it can be valuable info.

**Comment by paramat** [comment]
Since raillike depends on 18 neighbours how about creating a new direction table for face- and edge-connected and see how bad the bugs are with that? that would at least be some improvement.

**Comment by lhofhansl** [comment]
I wonder if this shows on my machine because I have a fast graphics card, so changes in CPU have a larger relative impact. Not sure. This one is not worth it generally it seems.

**Comment by lhofhansl** [comment]
@Fixer-007 I'll look through the diffs in these releases. Thanks for tracking these down!

**Comment by Fixer-007** [comment]
@lhofhansl I think I've nailed exact part: 
If you look here: https://github.com/minetest/minetest/blob/51cf464f7450ee31e3a986b521f90cc6195bb2fb/src/client.cpp#L42 , there is sleep_ms(50); that's why I had no stuttering on those versions, little later sleep time was decreased to 3ms, and from here I started to notice the stutter, and little later it was removed altogether in 2011, and since then I still have stuttering to this day. Just pasting back sleep_ms(15); or sleep_ms(20); removes the stutter caused by mesh updates... I can only hope if someone finds cause of jitter itself.

**Comment by Fixer-007** [comment]
>Ahh. It was FRAPS. Windows only, so I cannot use it here.

BTW, I think it is possible to output numbers from F5 graphs directly into debug.txt as somekind of csv output of all of them, that will be useful for everybody.

**Comment by lhofhansl** [comment]
Related to #5050

**Comment by lhofhansl** [comment]
@paramat I also like your idea of 18 directions. Since even 6 directions did not improve things for @Fixer-007 (still hard to believe) 18 for sure won't do anything.

**Comment by lhofhansl** [comment]
@Fixer-007 Note that the sleep_ms you mention is only happening when the queue is empty (q==NULL).
It's not same as your sleep (which happens each time we pop an item of the queue). I doubt that sleep (the lack thereof) is the culprit.

**Comment by numberZero** [comment]
@lhofhansl If the queue is empty, it should not `sleep`, it should `wait` instead. That is implemented now.
@Fixer-007 The problem might be that something locks mutexes too aggressively. That may be memory allocation, for example, it is used very extensively in the mesh update thread (`new` uses a mutex internally to be thread-safe AFAIK). Such problems are difficult to track, though, as they diminish on any debugging attempt (probably due to overall slowdown).

**Comment by lhofhansl** [comment]
@numberZero I get how it is working now. I was pointing out the difference to @Fixer-007 .

**Comment by lhofhansl** [comment]
But I too was suspecting the mutexes/semaphores used. Maybe it stutters on thread suspension/scheduling.
(good mutexes and semaphores spinlock for just a little bit and only then suspend the thread, I do not know what Linux pthreads are doing or whether this is done on Windows)

**Comment by Fixer-007** [comment]
Thats not only you who suspected this, also, @gregorycu also was trying to fix this, you can look his efforts here https://github.com/gregorycu/minetest/tree/Jitter Last time I tried I had no luck (but I don't remember it for sure).

**Comment by gregorycu** [comment]
I also saw this caused some jitter.

I considered making it so that the MeshGenerator thread copied the data, which means the main thread is blocked for less time and/or less frequently.

With regards to the sleeping, I can't recall, but is the number of meshes generated per game loop finite? For instance, if it will process 1 mesh every loop, but the duration between each loop is reduced, then more work will be done per second.

**Comment by lhofhansl** [comment]
@gregorycu Interesting. I tried that too.
[jitter-wip.txt](https://github.com/minetest/minetest/files/786589/jitter-wip.txt)

@Fixer-007 Tried that my patch (see #5050), but didn't see a great improvement. Also there's now a race condition; the blocks are still being modified in the main thread, which leads to crashes.

I saw a significant reduction in jitter with this (if only the client wouldn't crash now :) )

Is that the line of thinking you had in mind? Any ideas where I need to synchronize with the main thread in order to avoid problem - it looks like the location of blocks can change during the call MeshMakeData::fill (specifically between the point where there area for the voxel manipulator is sized and the point where the neighboring blocks are added)

**Comment by lhofhansl** [comment]
Just looked at your change. Much better than mine. I'll try that.

**Comment by gregorycu** [comment]
I'm not sure the code is good. Probably needs an expert eye to go over it. It's very much a WIP.

There are probably issues with it. For instance, I am not sure I've probably synchronized access to the block data, which is the thing that's probably causing your crash. (Sorry I can't be more help).

**Comment by lhofhansl** [comment]
Just ran with it a bit, leads to similar crashes as my naive change. :(
But I think we're on to something. The copying of the block data should not happen in the main thread.

**Comment by numberZero** [comment]
@lhofhansl The problem may arise e.g. if the main thread pushes many update requests in a row: the if generation thread starts working on the first request, it *will* race with the main thread in many places, including queue mutex, allocation mutex, etc. This is what a small sleep could solve.

Anyway, I will probably try to investigate a bit further. Don’t want to go down to disassembling, though...

**Comment by lhofhansl** [comment]
@numberZero yeah, unfortunately Minetest was not designed with rendering being asynchronous to changes to the map or blocks in mind. It's very hard to change this afterwards.

**Comment by lhofhansl** [comment]
I tried some other way to improve this. ATM I am stumped. Let's close this one. Perhaps I can open an issue and summarize and/or link to all the various efforts we have and had to improve the jitter.

**Comment by lhofhansl** [comment]
Actually lemme try with g18. If there're no visual artifact we have a slight improvement.
Currently when we add a block at the [client](https://github.com/minetest/minetest/blob/master/src/client.cpp#L1602), the block and its six neighbors are added, then we add the 26 neighbors all 7 times for a total of 189 blocks. With this this will down to 133 blocks at least.

(Note: I tried not adding the neighbors [here](https://github.com/minetest/minetest/blob/master/src/client.cpp#L1610) or [here](https://github.com/minetest/minetest/blob/master/src/mapblock_mesh.cpp#L86). In both cases I get rendering glitches.)

**Comment by lhofhansl** [comment]
Updated to use 18 neighbors instead. I can't spot any glitches with this, but then again others have been quite a lot better at that than I was.

**Comment by Fixer-007** [comment]
I will check again.

**Comment by lhofhansl** [comment]
OK then. Lemme close this.

**Comment by Fixer-007** [comment]
@lhofhansl I did further testing, it seems the shadow border glitch like this https://i.imgur.com/zMHJ44f.png also happens in clean game, I'm doing full recompile of newest version to be really sure. After I confirm this, this PR can be reopened again.

**Comment by numberZero** [comment]
@Fixer-007 it happens sometimes. I think that happens when a mesh is generated before its neighbors are loaded.

**Comment by Fixer-007** [comment]
@lhofhansl I did further testing, it seems the shadow border glitch like this https://i.imgur.com/zMHJ44f.png also happens in clean game.
**Please reopen this back** :)

**Comment by numberZero** [comment]
> Please reopen this back :)

@Fixer-007 certainly not this PR. Currently you can just dig or place any node (even protected) in the affected mapblock and the glitch goes away (from the neighbors too). With this, that would not help.

**Comment by numberZero** [comment]
@Fixer-007 actual solution would be to perform no copy at all; I have some ideas but that’s not straightforward due to multithreading issues.

**Comment by lhofhansl** [comment]
OK. Reopening.

**Comment by lhofhansl** [comment]
> @Fixer-007 certainly not this PR.

So what is it now? :)
This will reduce the work done by about 30%. Does it cause any glitches or not?
Of course it's not the final solution, but perhaps a step on the way.

No copy is better. Perhaps the block can the loaded straight into the voxel manipulator with its neighbors.
Doing the work not on the rendering thread is also better.

**Comment by Fixer-007** [comment]
I will play with it for few days.

**Comment by Fixer-007** [comment]
Have water bugs with this patch unfortunately:
default

**Comment by lhofhansl** [comment]
Thanks @Fixer-007 . Closing this one then.

**Pull request closed without merging**